Support lcc compiler for e2k (Elbrus) architecture

makise-homura opened the pull request:
Added a support for lcc compiler for Elbrus processors based on e2k architecture.
Examples of such CPUs:
https://en.wikipedia.org/wiki/Elbrus-8S
https://en.wikipedia.org/wiki/Elbrus-2S%2B

Such compiler have a similar behavior as gcc (option compatibility), but, instead of 'Free Software Foundation' string, its version infomation contains strings 'lcc' and 'e2k' like this:

lcc:1.21.22:Oct-15-2017:e2k-v4-linux
gcc (GCC) 4.8.0 compatible

So specific strings in environment.py were altered.

jpakkane commented:
Can you please pastebin the full output log of running `./run_unittests.py` with this compiler. Thanks.

makise-homura commented:
Well, it doesn't run on current master and says 'Meson works correctly only with python 3.5+.', and my setup has python 3.4.3.
(By the way, meson's requirements in README.md are still specifying it's 3.4 required.)
But the same fix is perfectly working on 0.44.0. If you're interested in it, I've attached log of ./run_unittests.py of meson 0.44.0.
Still I'll try to build python 3.5 for this machine and then re-run tests on master instead of 0.44.0 release.

By the way, not every unit test has passed (failures=2, errors=9, skipped=7). For example, I've noticed that lcc is not strictly option-compatible with gcc and has no -flto option in 1.21.22 (but, according to docs, it is announced in 1.23.01 and will be available when this compiler would be released); it does not support -std=c++17 (only limited of c++14); I have no valac in my setup (but possibly will be able to build and install it); etc, etc.
Here it is: https://pastebin.com/XN3xrcCt

makise-homura commented:
I've installed python 3.5 and vala 0.20 and ran unit tests on the branch I'm about to request pulling.
Here's the result: https://pastebin.com/yveZP2FJ
Still not perfect (failures=2, errors=9, skipped=7) of 87 tests.

Would you please check what's wrong with passing these tests?

I may notice from output that our current compiler does not support -fsanitize=address, -std=c++17, and -flto options; in addition, it has __builtin_hfkerhisadf and hfkerhisadf undefined (but what is it? I never spotted any of these identifiers), and it brings warning 'lchmod is not implemented and will always fail' instead of an error. Looks like that's not a meson issues, but compiler's ones. According to other failed tests, i don't know what to think about.

jpakkane commented:
Is the compiler meant to be a standalone thing (as indicated by its name)? If yes then we should probably make it be its own compiler object rather than being treated as stock GCC. The simplest way is to create a new compiler object that inherits from GnuCCompiler (and equiv for other languages) and add all extra stuff there and have options that match what the compiler can provide (such as not having an option for asan et al).

ignatenkobrain commented on mesonbuild/compilers/compilers.py:
```diff
@@ -1076,6 +1076,17 @@ def get_default_include_dirs(self):
+        self.id = 'lcc'
+        self.base_options = ['b_pgo', 'b_coverage',
+                             'b_ndebug', 'b_staticpic',
+                             'b_lundef', 'b_asneeded' ]
```
__[Flake8]__

[E202] whitespace before ']'


<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/compilers/cpp.py:
```diff
@@ -133,6 +134,21 @@ def get_pch_use_args(self, pch_dir, header):
+        opts['cpp_std'] = coredata.UserComboOption('cpp_std', 'C++ language standard to use',
+                                                   ['none', 'c++98', 'c++03', 'c++0x', 'c++11', 'c++14', 'c++1y',
+                                                    'gnu++98', 'gnu++03', 'gnu++0x', 'gnu++11', 'gnu++14', 'gnu++1y'],
+                                                    'none')
```
__[Flake8]__

[E127] continuation line over-indented for visual indent


<!-- Comment by [SideCI](https://sideci.com) -->

makise-homura commented:
Well, looks like I've done what you've proposed @jpakkane. Also added some fixes not directly related to lcc and python 3.5, but for issues that have arisen while I experimented with it.

Still unit tests do not finish without errors and failures: failures=2, errors=3, skipped=8.
Resulting log is here: https://pastebin.com/2WCPxpwX 

As far as I can see, these test were failed:
* `test_custom_target_changes_cause_rebuild`: I can't identify reason why.
* `test_source_changes_cause_rebuild`: I can't identify reason why.

And these were finished with 'error':
* `test_preprocessor_checks_CPPFLAGS`: I can't identify reason why.
* `test_pch_with_address_sanitizer`: error, due to lcc does not support bot PCH and sanitizers. I suppose this test should be skipped rather than executed in the case of lcc.
* `test_reconfigure`: says 'Problem encountered: b_lto not set', because LTO is not supported. I suppose another variable should be used there which is supported in every compiler, for example, 'b_pgo' or 'b_coverage'. 

Would you please direct me to what should I check in our compiler or meson to make first three tests pass, or figure out that these tests should be skipped, if so?

By the way, what's the difference between 'error' and 'failure' in a test?

And what's the decision of skipping some tests? Isn't it possible to skip `test_pch_with_address_sanitizer`, if we are using lcc?

jpakkane commented:
> Isn't it possible to skip test_pch_with_address_sanitizer, if we are using lcc?

Yes, by checking whether the compiler points to lcc (presumably by checking the CC envvar and checking if it points to lcc) and then throwing the test skip exception. Grep the file to see how it is used.

jpakkane commented:
The documentation also needs updating. First of all reference tables needs to be updated with the id information of this new compiler. Secondly a release note snippet needs to be added in the `docs/markdown/snippets` directory.

makise-homura commented:
Ok, I'll see it tomorrow. In addition, I'll see what I can do with remaining two unit tests that are failed: as long as I figured out, there's some strange behavior in deps parser of ninja (maybe, re2c works a bit incorrect), and supposedly it's not a meson problem. But when I'll be sure, I'll fix it and explain what was there.

makise-homura commented:
Well, I figured out why do the last two unit tests fail: our compiler incorrectly creates depencency files when paths contain spaces. It is no problem if we just build a project, but if we rebuilding it, ninja thinks that dependencies like '../../usr/src/meson/e2k/meson/test', 'cases/common/64', 'custom', 'header', and 'generator/prog.c' do not exist. I filed a bugreport there and will wait for new version of compiler; and until then, there's no way to let these tests pass.
So I noticed the exactly same behavior of Intel compiler by the code of common test case number 132; and implemented such skipping strategy into failing test cases number 64 and 22. And now unit tests are completely passed! Here's the output:
```
sakuya /usr/src/meson/e2k/meson # ./run_unittests.py
.........s....s...................s......s..s.........ss..s.....s.........s.....s......
----------------------------------------------------------------------
Ran 87 tests in 361.819s

OK (skipped=11)
sakuya /usr/src/meson/e2k/meson #
```

In addition, I updated reference tables and a release note snippet. I'm not finally sure, have I done everything correct with it, especially the snippet; feel free to correct it or ask me to do this if there's not everything is OK.

makise-homura commented on docs/markdown/Reference-tables.md:
```diff
@@ -42,6 +43,7 @@ set in the cross file.
 | x86                 | 32 bit x86 processor  |
 | x86_64              | 64 bit x86 processor  |
 | arm                 | 32 bit ARM processor |
+| e2k                 | MCST Elbrus processor |
```
Yes, it's the exactly same one that is available since automake 1.15.1.
About Python, I'm not sure (for example, in python 3.6.5 sources, I can't find any `e2k` substring in text files, except for automake-related `config.guess` and `config.sub`).

These processors are 64-bit from the beginning; but they support execution of 32-bit programs as well (provided that needed 32-bit libraries are supplied), for example, it is subject for non-portable programs that, say, treat unsigned int as uint32_t. It's nearly identical situation as with x86_64: it supports execution of both 64-bit and 32-bit code, even if OS kernel and CPU architecture is purely 64-bit (but with the exception that for x86_64, there is hardware 32-bit counterpart exist (x86); but there is no such counterpart for e2k).

jpakkane replied:
The detection is done with `platform.machine().lower()` in `detect_cpu_family`. Can you test that? If it returns the correct value, then everything should be fine. If not, please add the necessary code to return the correct value.

jpakkane replied:
Ping. We are approaching dev freeze for the next release and it would be great to get this in before that.

makise-homura replied:
I'll see this today or tomorrow (my elbrus testbench was busy in stability testing until yesterday's evening).

makise-homura replied:
As far as I see, it's all ok with this now:
```
The Meson build system
Version: 0.45.0.dev1
Source dir: /home/molchan_i/src/meson
Build dir: /home/molchan_i/src/meson/0
Build type: native build
Project name: simple
Native C compiler: cc (lcc 1.23.6 "lcc:1.23.06:Mar--4-2018:e2k-v4-linux")
Build machine cpu family: e2k
Build machine cpu: e8c
```
Note `e2k` in family and `e8c` in cpu (`e2k` was in both fields before commit 495f47b); I had to make more precise cpu detection due to the fact that e2k family has some different CPUs, like `e3m`, `e3s`, `e2c+`, `e2cv`, `e2c1`, `e2c1+`, `e8c`, `e12c`, and `e16c`.

makise-homura commented:
Well, I see AppVeyor build somehow shows as cancelled; I can't figure a reason why.
It failed last time, I suppose, due to the fact I tried to provide original names of CPUs in Russian (UTF-8) in snippet; I removed this, and pushed this commit, but now AppVeyor build shows to be cancelled.
Is there any problem from my side in this?

makise-homura commented:
In addition to cpu and cpu family stuff, I've also fixed version of compiler where correct work with depfiles is announced, so affected unit tests would be ran correctly on such new compiler versions rather than being skipped.
